external/fv3fit/fv3fit/keras/_models/models.py

```diff
@@ -36,10 +49,52 @@
 MODEL_DIRECTORY = "model_data"
 KERAS_CHECKPOINT_PATH = "model_checkpoints"
 
-# Description of the training loss progression over epochs
-# Outer array indexes epoch, inner array indexes batch (if applicable)
-EpochLossHistory = Sequence[Sequence[Union[float, int]]]
-History = Mapping[str, EpochLossHistory]
+
+@dataclasses.dataclass
+class DenseHyperparameters(Hyperparameters):
+    """
+    Configuration for training a dense neural network based model.
+
+    Args:
+        input_variables: names of variables to use as inputs
+        output_variables: names of variables to use as outputs
+        weights: loss function weights, defined as a dict whose keys are
+            variable names and values are either a scalar referring to the total
+            weight of the variable. Default is a total weight of 1
+            for each variable.
+        normalize_loss: if True (default), normalize outputs by their standard
+            deviation before computing the loss function
+        optimizer_config: selection of algorithm to be used in gradient descent
+        loss: loss function to use, should be 'mse' or 'mae'
+        save_model_checkpoints: if True, save one model per epoch when
+            dumping, under a 'model_checkpoints' subdirectory
+        nonnegative_outputs: if True, add a ReLU activation layer as the last layer
+            after output denormalization layer to ensure outputs are always >=0
+            Defaults to False.
+        fit_kwargs: other keyword arguments to be passed to the underlying
+            tf.keras.Model.fit() method
+    """
+
+    input_variables: List[str]
+    output_variables: List[str]
+    weights: Optional[Mapping[str, Union[int, float]]] = None
+    normalize_loss: bool = True
+    optimizer_config: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    dense_network: DenseNetworkConfig = dataclasses.field(
```

Comment:
Since this breaks current configs, can you raise an error with more specific info about this change that says to put the former top level hyperparameters under `dense_network, training_loop` if an old format config is provided?

Reply:
This is a really good point, but I don’t think I have time to add that today. I found out that if you use an old config, it actually raises no error, because we’re not using dacite in strict mode. Is this OK to either not add, or add in a PR where we fix dacite to use strict mode in fv3fit?